### PR TITLE
Add infer storage type function for sparse slice operator

### PR DESCRIPTION
**Pull request by anirudh2290**
This change will allow to fallback to dense operator for slice for features such as slicing along more than one axis which is not supported for sparse.

### CODE REVIEW

File: tests/python/unittest/test_sparse_ndarray.py [hunk 1]
```diff
@@ -113,9 +113,20 @@ def check_sparse_nd_csr_slice(shape):
         assert same(A[start:end].asnumpy(), A2[start:end])
         assert same(A[start:].asnumpy(), A2[start:])
         assert same(A[:end].asnumpy(), A2[:end])
+    
+    def check_slice_nd_csr_fallback(shape):
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_sparse_ndarray.py]: Could you also add
`from test_sparse_ndarray import test_sparse_nd_slice ` to `tests/python/gpu/test_operator_gpu.py` to make sure it fallbacks on gpu correctly?

File: src/operator/tensor/matrix_op-inl.h [hunk 2]
```diff
@@ -424,6 +424,41 @@ inline bool SliceShape(const nnvm::NodeAttrs& attrs,
   return true;
 }
 
+inline bool SliceForwardInferStorageType(const nnvm::NodeAttrs& attrs,
+                                           const int dev_mask,
```
[the remark this answers is not in the document]
Comment (anirudh2290) [reply src/operator/tensor/matrix_op-inl.h]: Done!

File: src/operator/tensor/matrix_op-inl.h [hunk 3]
```diff
@@ -424,6 +424,41 @@ inline bool SliceShape(const nnvm::NodeAttrs& attrs,
...
+                                           DispatchMode* dispatch_mode,
+                                           std::vector<int>* in_attrs,
+                                           std::vector<int>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 1);
+  CHECK_EQ(out_attrs->size(), 1);
+  const SliceParam& param = nnvm::get<SliceParam>(attrs.parsed);
+  auto& in_stype = in_attrs->at(0);
```
[the remark this answers is not in the document]
Comment (anirudh2290) [reply src/operator/tensor/matrix_op-inl.h]: Done.

File: tests/python/unittest/test_sparse_ndarray.py (hunk 1 above)
Comment (anirudh2290) [reply tests/python/unittest/test_sparse_ndarray.py]: Added.

File: tests/python/unittest/test_sparse_ndarray.py [hunk 4]
```diff
@@ -113,9 +113,20 @@ def check_sparse_nd_csr_slice(shape):
...
+        stype = 'csr'
+        A, _ = rand_sparse_ndarray(shape, stype)
+        A2 = A.asnumpy()
+        start = rnd.randint(0, shape[0] - 1)
+        end = rnd.randint(start + 1, shape[0])
+        result = mx.nd.sparse.slice(A, begin=(start, shape[1] - 1), end=(end + 1, shape[1]))
```
Comment (eric-haibin-lin) [inline tests/python/unittest/test_sparse_ndarray.py]: is begin on axis=0 and end on axis=1?

Comment (anirudh2290) [reply tests/python/unittest/test_sparse_ndarray.py]: begin means the starting coordinate on one or more axis. The position of values in the tuple indicate which axis to slice along For example for a 3d matrix of shape(3,3,3) slicing with begin=(0,0,0) and end=(2,2,2) will return a 2*2*2 matrix.

Comment (eric-haibin-lin) [reply tests/python/unittest/test_sparse_ndarray.py]: Got it

**Pull request merged**